031-custom_authentication.md

```diff
@@ -0,0 +1,198 @@
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=999999999
+listener.name.<listener-name>.sasl.enabled.mechanisms=oauthbearer
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+principal.builder.class=SimplePrincipal.class
+listener.security.protocol.map=CONTROLPLANE-9090:SSL,REPLICATION-9091:SSL,TLS-9093:SASL_SSL
+```
+
+To achieve this, we'd simply add the following listener config to our Kafka custom resource.
+
+```
+    listeners:
+      - name: custom-auth-listener
+        port: 9093
+        type: internal
+        tls: true
+        configuration:
+          brokerCertChainAndKey:
+            secretName: my-secret
+            certificate: my-certifciate.crt
+            key: my-key.key
+        authentication:
+          type: custom
+          sasl: true
+          principal.builder.class: SimplePrincipal.class
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
+            sasl.enabled.mechanisms: oauthbearer
+            oauthbearer.sasl.jaas.config: |
+              org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+          tlsTrustedCertificates: ...
+          secrets:
+            - name: example
+              mountPath: "/etc/example"
+```
+
+Then, when constructing the broker config, we’ll perform the following tasks:
+
+* `Principal`, if set, is set cluster-wide for all authentication methods. This is a limitation of Kafka, which only allows one principal to be specified for the entire cluster. If set, we need to ensure that no other listeners override this property, and if they do and are different, then fail-hard.
+* The protocol for this listener would be derived from the `tls` and `sasl`, which then would be appended to `listener.security.protocol.map`. For example, if `tls: true` and `sasl: true`, then the protocol will be `SASL_SSL`. 
+* Each configuration entry under `listener-config` would be pre-appended with `listener.name.<listener-name>`. 
+* `brokerCertChainAndKey` can be specified if a custom certificate wishes to be served from this listener. This automatically creates/shoves the specified certifciate into a keystore automatically for the user, as it's done for other authentication mechanisms today.
+* `tlsTrustedCertificates` functions identically to OAuth’s setting. This is needed for workflows which will use custom mTLS, or mTLS + SASL workflows, as they may require to add additional certificate authorities. Similar to `brokerCertChainAndKey`, a truststore is generated automatically for the user, with their provided certifciate authorities. 
```

Comment:
I think this part is problematic:
```
a truststore is generated automatically for the user
```

How does the system know what is truststore and what is keystore (important for automatically creating it). I think to make this versatile, this maybe needs to be merged with the `secrets` field just as opaque list of secrets which will be mounted and the user has to deal with them. If the authentication needs truststore or keystore, it might need to be already in the secret.

Reply:
I'm totally fine with just requiring the end user with creating their truststores/keystores. What I was thinking here was  just leveraging what's already in place: https://github.com/strimzi/strimzi-kafka-operator/blob/main/docker-images/kafka/scripts/kafka_tls_prepare_certificates.sh#L79 & https://github.com/strimzi/strimzi-kafka-operator/blob/79e28149b6dd8dd7b61e0ed60aae6d26fb93b1b7/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L200-L206 & https://github.com/strimzi/strimzi-kafka-operator/blob/79e28149b6dd8dd7b61e0ed60aae6d26fb93b1b7/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java#L1482